### PR TITLE
Twilio multipart_count

### PR DESCRIPTION
**Pull request by nickpell**
Fetch `multipart_count` for twilio messages directly via the API

TODO:
- [x] Rebase when https://github.com/dimagi/commcare-hq/pull/11385 is merged
- [x] Add tests

### CODE REVIEW

File: corehq/apps/smsbillables/models.py [hunk 1]
```diff
@@ -334,6 +333,9 @@ def create(cls, message_log, multipart_count=1):
         billable.gateway_fee = gateway_charge_info.gateway_fee
         billable.gateway_fee_conversion_rate = gateway_charge_info.conversion_rate
         billable.direct_gateway_fee = gateway_charge_info.direct_gateway_fee
+        billable.multipart_count=cls._get_multipart_count(
```
Comment (dimagimon) [inline corehq/apps/smsbillables/models.py]: E225 missing whitespace around operator

**Comment by nickpell** [comment]
@gcapalbo

File: corehq/apps/smsbillables/models.py [hunk 2]
```diff
@@ -343,6 +345,13 @@ def create(cls, message_log, multipart_count=1):
         return billable
 
     @classmethod
+    def _get_multipart_count(cls, backend_api_id, backend_instance, backend_message_id, multipart_count):
+        if backend_api_id == SQLTwilioBackend.get_api_id():
+            return int(get_twilio_message(backend_instance, backend_message_id).num_segments)
```
Comment (gcapalbo) [inline corehq/apps/smsbillables/models.py]: would be nice to only make one call to get the message from twilio per each billable request

**Comment by nickpell** [comment]
@gcapalbo added caching for twilio message objects (with timeout < retry delay)

File: corehq/apps/smsbillables/models.py [hunk 3]
```diff
@@ -347,7 +347,15 @@ def create(cls, message_log, multipart_count=1):
...
     def _get_multipart_count(cls, backend_api_id, backend_instance, backend_message_id, multipart_count):
         if backend_api_id == SQLTwilioBackend.get_api_id():
-            return int(get_twilio_message(backend_instance, backend_message_id).num_segments)
+            twilio_message = get_twilio_message(backend_instance, backend_message_id)
+            if twilio_message.num_segments is not None:
+                return int(twilio_message.num_segments)
+            else:
+                log_smsbillables_error(
+                    "num_segments is None: backend_message_id=%s, status=%s"
+                    % (twilio_message.backend_message_id, twilio_message.status)
```
Comment (gcapalbo) [inline corehq/apps/smsbillables/models.py]: this should just reference the local backend_message_id, right? i don't think backend_message_id exists on twilio_message

**Comment by gcapalbo** [comment]
@nickpell just one comment left otherwise looks good

File: corehq/apps/smsbillables/models.py (hunk 3 above)
Comment (nickpell) [inline corehq/apps/smsbillables/models.py]: updated

**Comment by nickpell** [comment]
✅

**Pull request merged**